presto-main/src/main/java/com/facebook/presto/operator/SortedPositionLinks.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+public final class SortedPositionLinks
+        implements PositionLinks
+{
+    public static class Builder implements PositionLinks.Builder
+    {
+        private final Map<Integer, List<Integer>> positionLinks;
+        private final int size;
+        private final Comparator<Integer> comparator;
+        private JoinFilterFunction filterFunction;
+
+        public Builder(int size, Comparator<Integer> comparator)
+        {
+            this.comparator = comparator;
+            this.size = size;
+            positionLinks = new HashMap<>();
+        }
+
+        @Override
+        public boolean link(int from, int to)
```

Comment:
yes, head of the position links must be the smallest value. With your proposed change to return `int` instead of boolean and with one comment that I have added, I think that this code is now easier to understand.